pkg/controller/ingressstate/ingress_state_controller.go

```diff
@@ -0,0 +1,328 @@
+type checkEndpointsFunc func(endpointIP string) error
+
+// degradedIngressConditions checks the health of a set of endpoints
+// and returns degraded operator conditions if any of the endpoints or
+// the pods that back them are unhealthy.
+func degradedIngressConditions(endpoints *corev1.Endpoints, checkPod checkPodFunc, checkEndpoints checkEndpointsFunc) ([]operatorv1.OperatorCondition, error) {
+	if endpoints == nil {
+		return []operatorv1.OperatorCondition{
+			endpointsDegraded("MissingEndpoints", "No endpoints found for oauth-server"),
+		}, nil
+	}
+
+	if len(endpoints.Subsets) == 0 {
+		return []operatorv1.OperatorCondition{
+			endpointsDegraded("ZeroEndpointsSubsets", "No endpoints found for oauth-server"),
+		}, nil
+	}
+
+	conditions := []operatorv1.OperatorCondition{}
+
+	// Aggregate unhealthy endpoints and pods across all subsets to
+	// support going degraded only if more than one endpoint or pod is
+	// unhealthy. A single endpoint or pod would commonly be unhealthy
+	// during an upgrade, and it is not desirable to report degraded
+	// in that case.
+	unhealthyEndpoints := []string{}
+	unhealthyPodCount := 0
+	unhealthyPodMessages := []string{}
+
+	for _, subset := range endpoints.Subsets {
+		// Report degraded if all addresses for the endpoint are not ready
+		if len(subset.Addresses) == 0 && len(subset.NotReadyAddresses) > 0 {
+			msg := fmt.Sprintf("%d endpoints for oauth-server are reporting 'not ready'", len(subset.NotReadyAddresses))
+			conditions = append(conditions, endpointsDegraded("NonReadyEndpoints", msg))
+			continue
+		}
+
+		for _, address := range subset.Addresses {
+			if err := checkEndpoints(address.IP); err != nil {
+				unhealthyEndpoints = append(unhealthyEndpoints, fmt.Sprintf("%s:%v", address.IP, err))
+			}
+
+			if address.TargetRef != nil && address.TargetRef.Kind == "Pod" {
+				podMessages := checkPod(address.TargetRef)
+				if len(podMessages) > 0 {
+					unhealthyPodCount += 1
+					unhealthyPodMessages = append(unhealthyPodMessages, podMessages...)
+				}
+			}
+		}
+	}
+
+	// Tolerate a single unhealthy endpoint to allow for termination during upgrade
+	if len(unhealthyEndpoints) > 1 {
+		msg := fmt.Sprintf("Unhealthy endpoints found: %s", strings.Join(unhealthyEndpoints, ","))
+		conditions = append(conditions, endpointsDegraded("UnhealthyEndpoints", msg))
+	}
+
+	// Tolerate a single unhealthy pod to allow for termination during upgrade
+	if unhealthyPodCount > 1 {
+		conditions = append(conditions, operatorv1.OperatorCondition{
+			Type:    "IngressStatePodsDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "UnhealthyPods",
+			Message: fmt.Sprintf("Unhealthy pods found: %s", strings.Join(unhealthyPodMessages, ",")),
+		})
+	}
+
+	// Ensure condition names are found in degradedConditionNames
+	for _, condition := range conditions {
+		if !degradedConditionNames.Has(condition.Type) {
+			return nil, fmt.Errorf("Unsupported degraded condition name %q", condition.Type)
+		}
+	}
```

Comment:
Why, don't you trust yourself? Isn't this more of a unit-test material rather than production code?

Reply:
Moved to unit test.